packages/xforms-engine/src/lib/reactivity/createInstanceValueState.ts

```diff
@@ -94,39 +94,46 @@ const guardDownstreamReadonlyWrites = (
 	return [getValue, setValue];
 };
 
-/**
- * Per {@link https://getodk.github.io/xforms-spec/#preload-attributes:~:text=concatenation%20of%20%E2%80%98uuid%3A%E2%80%99%20and%20uuid()}
- */
-const PRELOAD_UID_EXPRESSION = 'concat("uuid:", uuid())';
-
 /**
  * @todo It feels increasingly awkward to keep piling up preload stuff here, but it won't stay that way for long. In the meantime, this seems like the best way to express the cases where `preload="uid"` should be effective, i.e.:
  *
  * - When an instance is first loaded ({@link isInstanceFirstLoad})
  * - When an instance is initially loaded for editing ({@link isEditInitialLoad})
  */
-const shouldPreloadUID = (context: ValueContext) => {
+const isLoading = (context: ValueContext) => {
 	return isInstanceFirstLoad(context) || isEditInitialLoad(context);
 };
 
-/**
- * @todo This is a temporary one-off, until we support the full range of
- * {@link https://getodk.github.io/xforms-spec/#preload-attributes | preloads}.
- */
-const setPreloadUIDValue = (context: ValueContext, valueState: RelevantValueState): void => {
-	const { preload } = context.definition.bind;
+const postloadValue = (
+	context: ValueContext,
+	setValue: SimpleAtomicStateSetter<string>,
+	preload: AnyBindPreloadDefinition
+) => {
+	context.definition.model.registerXformsRevalidateListener(() => {
+		const value = preload.getValue(context);
+		if (value) {
+			setValue(value);
+		}
+	});
+};
 
-	if (preload?.type !== 'uid' || !shouldPreloadUID(context)) {
+const preloadValue = (context: ValueContext, setValue: SimpleAtomicStateSetter<string>): void => {
+	const { preload } = context.definition.bind;
+	if (!preload) {
 		return;
 	}
 
-	const preloadUIDValue = context.evaluator.evaluateString(PRELOAD_UID_EXPRESSION, {
-		contextNode: context.contextNode,
-	});
-
-	const [, setValue] = valueState;
+	if (preload.event === XFORM_EVENT.xformsRevalidate) {
+		postloadValue(context, setValue, preload);
+		return;
+	}
 
-	setValue(preloadUIDValue);
+	if (isLoading(context)) {
+		const value = preload.getValue(context);
```

Comment:
Is this the same code as in the listener callback (L113)? You might want to consider extracting it into a function, just to help us remember to update all those places consistently.